pulse2percept/retina.py

```diff
@@ -1001,63 +1425,53 @@ def jansonius(num_cells=500, num_samples=801, center=np.array([15, 2]),
+@utils.deprecated(deprecated_version='0.3', removed_version='0.4')
+def make_axon_map(xg, yg, jan_x, jan_y, axon_lambda=1, min_weight=0.001):
     """Retinal axon map
 
     Generates a mapping of how each pixel in the retina space is affected
     by stimulation of underlying ganglion cell axons.
     Parameters
     ----------
-    xg, yg : array
+    xg, yg: array
         meshgrid of pixel locations in units of visual angle sp
-    axon_lambda : float
-        space constant for how effective stimulation (or 'weight') falls off
+    axon_lambda: float
+        space constant for how effective stimulation(or 'weight') falls off
         with distance from the pixel back along the axon toward the optic disc
         (default 1 degree)
-    min_weight : float
+    min_weight: float
         minimum weight falloff.  default .001
 
     Returns
     -------
-    axon_id : list
-        a list, for every pixel, of the index into the pixel in xg,yg space,
+    axon_id: list
+        a list, for every pixel, of the index into the pixel in xg, yg space,
         along the underlying axonal pathway.
-    axon_weight : list
-        a list, for every pixel, of the axon weight into the pixel in xg,yg
+    axon_weight: list
+        a list, for every pixel, of the axon weight into the pixel in xg, yg
         space
 
     """
-    # initialize tuples
-    axon_xg = ()
-    axon_yg = ()
-    axon_dist = ()
-    axon_weight = ()
-    axon_id = ()
-
-    # loop through pixels as indexed into a single dimension
-    for px in range(0, len(xg.flat)):
+    axon_id = []
+    axon_weight = []
+    for idx, _ in enumerate(xg.ravel()):
+        cur_xg = xg.ravel()[idx]
+        cur_yg = yg.ravel()[idx]
         # find the nearest axon to this pixel
-        d = (jan_x - xg.flat[px])**2 + (jan_y - yg.flat[px])**2
+        d = (jan_x - cur_xg) ** 2 + (jan_y - cur_yg) ** 2
         cur_ax_id = np.nanargmin(d)  # index into the current axon
+
+        # `ax_num`: which axon it is
+        # `ax_pos_id0`: the point on that axon that is closest to `px`
         [ax_pos_id0, ax_num] = np.unravel_index(cur_ax_id, d.shape)
 
         dist = 0
-
-        cur_xg = xg.flat[px]
-        cur_yg = yg.flat[px]
-
-        # add first values to the list for this pixel
-        axon_dist = axon_dist + ([0],)
-        axon_weight = axon_weight + ([1],)
-        axon_xg = axon_xg + ([cur_xg],)
-        axon_yg = axon_yg + ([cur_yg],)
-        axon_id = axon_id + ([px],)
-
-        # now loop back along this nearest axon toward the optic disc
+        this_id = [idx]
+        this_weight = [1.0]
         for ax_pos_id in range(ax_pos_id0 - 1, -1, -1):
             # increment the distance from the starting point
-            ax = (jan_x[ax_pos_id + 1, ax_num] - jan_x[ax_pos_id, ax_num])**2
-            ay = (jan_y[ax_pos_id + 1, ax_num] - jan_y[ax_pos_id, ax_num])**2
+            ax = (jan_x[ax_pos_id + 1, ax_num] - jan_x[ax_pos_id, ax_num])
```

Comment:
This fixes the bug. Could we add a test that checks this, and prevents us from ever putting it back in? (maybe you have that? I couldn't tell...)

Reply:
I did not test it in `make_axon_map`, which we now declare deprecated. I do have a test that makes sure the distance calculation is right: see `test_retina.py:test_axon_dist_from_soma`.